contracts/adapters/swap/dexter/src/contract.rs

```diff
@@ -0,0 +1,504 @@
+
+    let dexter_vault_contract_address = deps.api.addr_validate(&msg.dexter_vault_contract_address)?;
+    let dexter_router_contract_address = deps.api.addr_validate(&msg.dexter_router_contract_address)?;
+
+    // Store the entry point contract address
+    ENTRY_POINT_CONTRACT_ADDRESS.save(deps.storage, &checked_entry_point_contract_address)?;
+    DEXTER_ROUTER_ADDRESS.save(deps.storage, &dexter_router_contract_address)?;
+    DEXTER_VAULT_ADDRESS.save(deps.storage, &dexter_vault_contract_address)?;
+
+    Ok(Response::new()
+        .add_attribute("action", "instantiate")
+        .add_attribute(
+            "entry_point_contract_address",
+            checked_entry_point_contract_address.to_string(),
+        ))
+}
+
+///////////////
+/// RECEIVE ///
+///////////////
+
+// Receive is the main entry point for the contract to
+// receive cw20 tokens and execute the swap
+pub fn receive_cw20(
+    deps: DepsMut,
+    env: Env,
+    mut info: MessageInfo,
+    cw20_msg: Cw20ReceiveMsg,
+) -> ContractResult<Response> {
+    let sent_asset = Asset::Cw20(Cw20Coin {
+        address: info.sender.to_string(),
+        amount: cw20_msg.amount,
+    });
+    sent_asset.validate(&deps, &env, &info)?;
+
+    // Set the sender to the originating address that triggered the cw20 send call
+    // This is later validated / enforced to be the entry point contract address
+    info.sender = deps.api.addr_validate(&cw20_msg.sender)?;
+
+    match from_json(&cw20_msg.msg)? {
+        Cw20HookMsg::Swap { operations } => execute_swap(deps, env, info, operations),
+    }
+}
+
+///////////////
+/// EXECUTE ///
+///////////////
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> ContractResult<Response> {
+    match msg {
+        ExecuteMsg::Receive(cw20_msg) => receive_cw20(deps, env, info, cw20_msg),
+        ExecuteMsg::Swap { operations } => {
+            one_coin(&info)?;
+            execute_swap(deps, env, info, operations)
+        }
+        ExecuteMsg::TransferFundsBack {
+            swapper,
+            return_denom,
+        } => Ok(execute_transfer_funds_back(
+            deps,
+            env,
+            info,
+            swapper,
+            return_denom,
+        )?),
+        _ => {
+            panic!("NOT IMPLEMENTED");
+        }
+    }
+}
+
+fn execute_swap(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    operations: Vec<SwapOperation>,
+) -> ContractResult<Response> {
+    // Get entry point contract address from storage
+    let entry_point_contract_address = ENTRY_POINT_CONTRACT_ADDRESS.load(deps.storage)?;
+    let dexter_router_contract_address = DEXTER_ROUTER_ADDRESS.load(deps.storage)?;
+
+    // Enforce the caller is the entry point contract
+    if info.sender != entry_point_contract_address {
+        return Err(ContractError::Unauthorized);
+    }
+
+    let coin_in = one_coin(&info)?;
```

Comment:
Does dexter have cw20 support / expect it to have it? If so this one_coin verification would result in an error for cw20 swaps since there would be no coin attached to the info

Reply:
Makes sense. Thought currently we don't have any CW20 pools, but we can make this generic.